api/v1alpha7/types.go

```diff
@@ -164,10 +164,25 @@ type RootVolume struct {
 }
 
 type AdditionalBlockDevice struct {
-	// Name of the Cinder volume in the context of a machine.
-	// It will be combined with the machine name to make the actual volume name.
+	// BlockDeviceType is the type of block device to create.
+	// This can be either "volume" or "local".
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=volume;local
+	BlockDeviceType string `json:"blockDeviceType"`
```

Comment:
What we've got here is a discriminated union.

BlockDeviceType = `volume`:
* `VolumeType`
* `AvailabilityZone`

BlockDeviceType = `local`:
* `GuestFormat`

Please can you find a few examples of prior art in the k8s api so we can consider the pros and cons of existing designs?

Reply:
This looks like it might be interesting reading: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md